pkg/controllers/provisioning/scheduling/scheduler.go

```diff
@@ -50,61 +51,106 @@ func init() {
 }
 
 type Scheduler struct {
-	kubeClient client.Client
-	topology   *Topology
+	kubeClient  client.Client
+	preferences *Preferences
 }
 
 func NewScheduler(kubeClient client.Client) *Scheduler {
 	return &Scheduler{
-		kubeClient: kubeClient,
-		topology:   &Topology{kubeClient: kubeClient},
+		kubeClient:  kubeClient,
+		preferences: NewPreferences(),
 	}
 }
 
-func (s *Scheduler) Solve(ctx context.Context, provisioner *v1alpha5.Provisioner, instanceTypes []cloudprovider.InstanceType, pods []*v1.Pod) ([]*Node, error) {
+func (s *Scheduler) Solve(ctx context.Context, constraints *v1alpha5.Constraints, instanceTypes []cloudprovider.InstanceType, pods []*v1.Pod) ([]*Node, error) {
 	defer metrics.Measure(schedulingDuration.WithLabelValues(injection.GetNamespacedName(ctx).Name))()
-	constraints := provisioner.Spec.Constraints.DeepCopy()
 
 	sort.Slice(pods, byCPUAndMemoryDescending(pods))
 	sort.Slice(instanceTypes, byPrice(instanceTypes))
 
-	// Inject temporarily adds specific NodeSelectors to pods, which are then
-	// used by scheduling logic. This isn't strictly necessary, but is a useful
-	// trick to avoid passing topology decisions through the scheduling code. It
-	// lets us treat TopologySpreadConstraints as just-in-time NodeSelectors.
-	if err := s.topology.Inject(ctx, constraints, pods); err != nil {
-		return nil, fmt.Errorf("injecting topology, %w", err)
+	topology := NewTopology(s.kubeClient, &constraints.Requirements)
+	if err := topology.Initialize(ctx, pods...); err != nil {
+		return nil, fmt.Errorf("tracking topology counts, %w", err)
 	}
-
-	nodeSet, err := NewNodeSet(ctx, constraints, s.kubeClient)
+	daemonOverhead, err := s.getDaemonOverhead(ctx, constraints)
 	if err != nil {
-		return nil, fmt.Errorf("constructing nodeset, %w", err)
+		return nil, err
```

Comment:
Worth doing a pass on all of these errors at some point to make sure they're annotated where useful.